### DISCUSSION TITLE
Too many open files

### DISCUSSION THREAD

**Discussion by gbook** [rikyoz/bit7z#209] [General]
I'm experiencing an intermittent error. When using the following function:
````
bool squirrel::CompressDirectoryToArchive(QString dir, QString archivePath, QString &m) {
    utils::Print("Inside CompressDirectoryToArchive()");

    try {
        using namespace bit7z;
#ifdef Q_OS_WINDOWS
        Bit7zLibrary lib("C:/Program Files/7-Zip/7z.dll");
#else
        Bit7zLibrary lib("/usr/libexec/p7zip/7z.so");
#endif
        if (archivePath.endsWith(".zip", Qt::CaseInsensitive)) {
            BitArchiveWriter archive(lib, BitFormat::Zip);
            archive.setUpdateMode(UpdateMode::Update);
            archive.addDirectory(dir.toStdString());
            archive.compressTo(archivePath.toStdString());
        }
        else {
            BitArchiveWriter archive(lib, BitFormat::SevenZip);
            archive.setUpdateMode(UpdateMode::Update);
            archive.addDirectory(dir.toStdString());
            archive.compressTo(archivePath.toStdString());
        }
        m = "Successfully compressed directory [" + dir + "] to archive [" + archivePath + "]";
        return true;
    }
    catch ( const bit7z::BitException& ex ) {
        /* Do something with ex.what()...*/
        m = "Unable to compress directory into archive using bit7z library [" + QString(ex.what()) + "]";
        return false;
    }
}
````

... sometimes it works correctly and writes out the 7zip file. Sometimes it fails with bit7z giving this error from `ex.what()`:

`Failed to open the output file: Too many open files`

What might be happening?

**Comment by rikyoz** [comment]
This is weird. What operating system is this happening on?

**Comment by gbook** [comment]
RHEL8

**Comment by rikyoz** [comment]
As far as I remember, bit7z doesn't keep many opened files.
The only exception is when creating a multi-volume archive, but this is not your case.
Are there many files inside the directory you're compressing?
I'm not sure this an issue with bit7z, but I'll investigate it!

**Comment by gbook** [comment]
It's intermittent. I've run `lsof` while running my program with bit7z, and the number of open file handles is minmal. Maybe 50 or so open files, but way less than the OS limit of 1024.

**Reply by rikyoz** [reply]
Really strange. I ran some tests in my Fedora 28 VM but couldn't reproduce the problem.

In my tests, I used the strace command to trace the calls to the `open`, `openat` and `close` syscalls performed by the same test program I created to diagnose issue #206.
The program always closed the files it opened, and never kept too many files open at the same time.

Unfortunately, the fact that it's intermittent doesn't make it easy to replicate.

**Reply by gbook** [reply]
I'll keep trying it on my end and see what I can find.
